test/solver/solver.cpp

```diff
@@ -0,0 +1,675 @@
+        };
+        {
+            SCOPED_TRACE("Empty matrix (0x0)");
+            guarded_fn(gen_mtx(0, 0, 0, 0));
+        }
+        {
+            SCOPED_TRACE("Sparse Matrix with variable row nnz (50x50)");
+            guarded_fn(gen_mtx(50, 50, 10, 20));
+        }
+    }
+
+    template <typename TestFunction>
+    void forall_solver_scenarios(const test_pair<Mtx>& mtx, TestFunction fn)
+    {
+        auto guarded_fn = [&](auto solver) {
+            try {
+                fn(std::move(solver));
+                this->reset_rand();
+            } catch (std::exception& e) {
+                FAIL() << e.what();
+            }
+        };
+        {
+            SCOPED_TRACE("Unpreconditioned solver with 0 iterations");
+            guarded_fn(test_pair<SolverType>{
+                Config::build(ref, 0).on(ref)->generate(mtx.ref),
+                Config::build(exec, 0).on(exec)->generate(mtx.dev)});
+        }
+        if (Config::is_preconditionable()) {
+            SCOPED_TRACE("Preconditioned solver with 0 iterations");
+            guarded_fn(test_pair<SolverType>{
+                Config::build_preconditioned(ref, 0).on(ref)->generate(mtx.ref),
+                Config::build_preconditioned(exec, 0).on(exec)->generate(
+                    mtx.dev)});
+        }
+        if (Config::is_iterative()) {
+            {
+                SCOPED_TRACE("Unpreconditioned solver with 4 iterations");
+                guarded_fn(test_pair<SolverType>{
+                    Config::build(ref, 4).on(ref)->generate(mtx.ref),
+                    Config::build(exec, 4).on(exec)->generate(mtx.dev)});
+            }
+            if (Config::is_preconditionable()) {
+                SCOPED_TRACE("Preconditioned solver with 4 iterations");
+                guarded_fn(test_pair<SolverType>{
+                    Config::build_preconditioned(ref, 4).on(ref)->generate(
+                        mtx.ref),
+                    Config::build_preconditioned(exec, 4).on(exec)->generate(
+                        mtx.dev)});
+            }
+        }
+    }
+
+    template <typename VecType = Vec, typename TestFunction>
+    void forall_vector_scenarios(const test_pair<SolverType>& solver,
+                                 TestFunction fn)
+    {
+        auto guarded_fn = [&](auto b, auto x) {
+            try {
+                fn(std::move(b), std::move(x));
+                this->reset_rand();
+            } catch (std::exception& e) {
+                FAIL() << e.what();
+            }
+        };
+        {
+            SCOPED_TRACE("Multivector with 0 columns");
+            guarded_fn(gen_in_vec<VecType>(solver, 0, 0),
+                       gen_out_vec<VecType>(solver, 0, 0));
+        }
+        {
+            SCOPED_TRACE("Single vector");
+            guarded_fn(gen_in_vec<VecType>(solver, 1, 1),
+                       gen_out_vec<VecType>(solver, 1, 1));
+        }
+        // GMRES and others are currently broken w.r.t lucky breakdown
```

Comment:
Also an issue, except if fixed somewhere.

Reply:
Not yet, so I'll have to add a issue/fix for that as well